### DISCUSSION TITLE
Allow Generated Mocks to Accept Other Mock Interfaces as Input to Avoid Cyclic Imports

### DISCUSSION THREAD

**Discussion by kAvEh--** [uber-go/mock#212] [Q&A] [answered]
*Requested feature* 
Consider two interfaces, `InterfaceA` and `InterfaceB`. `InterfaceA` has a function that takes an object of `InterfaceB` as input. I want to generate `MockInterfaceA`, with the function accepting `MockInterfaceB` as input instead of `InterfaceB`.

*Why the feature is needed* 
If the generated mock files are placed in a separate folder, and the mocked interface is used in a test file located in the main package, it will result in a cyclic import: main package -> mock package -> main package.

**Comment by tulzke** [comment]
Here is a simple example why this does not work:
[example.zip](https://github.com/user-attachments/files/17234409/example.zip)

TL;DR:
// compilation error
// Type HumanMock does not implement Human
// need the method: Call(dog Dog)
// have the method: Call(_ DogMock)
var human Human = mock.HumanMock{}

**Comment by tulzke** [comment, marked as answer]
Possible solutions to the problem:
1. Generate mocks to the same directory as interfaces.
2. Use the `_test` postfix for package in tests. Then the dependency graph will look like this:
```mermaid
flowchart LR

some_package
mock --> some_package
some_package_test --> some_package
some_package_test --> mock
```
